### PR TITLE
Fix DOTD script to use branch names, not commit hashes.

### PR DESCRIPTION
**Pull request by ashercodeorg**

### CODE REVIEW

**Comment by ashercodeorg** [comment]
When reviewing, please think about whether this is robust for all reasonable initial conditions.

**Comment by ashercodeorg** [comment]
This (admittedly significant) change was made in #14880 to help unblock continuous DTTs. The thought (discussed offline) was that it would become an increasingly common situation to have the next (probably automated) DTT already in progress when time to DTP. Thus, to account for this, the script was changed to DTP by commit in #14880 rather than by branch.

Unfortunately, though the Octokit (GitHub) API often allows both commit or branch in its API, here it does not. Thus, the change in #14880 was bad. This PR attempts to fix it.

File: bin/dotd [hunk 1]
```diff
@@ -348,39 +348,56 @@ def main
...
 
-        should_i 'commit content changes from levelbuilder, to help prevent merge conflicts' do
-          content_push 'levelbuilder-staging', 'levelbuilder'
+      # As GitHub's API (all of ours, Octokit's and github.com's) does not allow creating a pull
+      # request from a commit hash (only a branch name), we create and push a branch.
+      branch_name = "dtp_candidate_#{test_green_commit}"
+      begin
+        run_on(
+          'test',
+          "cd test && git checkout test && git checkout -b #{branch_name} #{test_green_commit} && git push origin #{branch_name}"
```
Comment (balderdash) [inline bin/dotd:358]: I think that last part needs to be `git push origin test:#{branch_name}`

File: bin/dotd [hunk 2]
```diff
@@ -348,39 +348,56 @@ def main
...
-        DevelopersTopic.set_dtp 'yes'
-      end
-      should_i 'update Slack#infra-production topic' do
-        sha = GitHub.sha 'production'
-        InfraProductionTopic.set_dtp_commit(sha)
+        puts_after_dtp_tasks
+      ensure
+        run_on(
+          'test',
+          "cd test && git checkout test && git branch -D #{branch_name}"
```
Comment (balderdash) [inline bin/dotd]: Can you delete the remote branch here as well?

File: bin/dotd (hunk 1 above)
Comment (aoby) [reply bin/dotd:358]: Wouldn't that push the head of test to the remote branch? Here we want to push the current active branch, which this command will do AFAICT

File: bin/dotd (hunk 2 above)
Comment (aoby) [reply bin/dotd]: @balderdash what's the advantage of reusing the same branch? I don't actually have a strong preference here, just curious. Git branches are lightweight.

File: bin/dotd (hunk 1 above)
Comment (balderdash) [reply bin/dotd:358]: Oh right, it checks out `branch_name` first. Yeah this is fine.

File: bin/dotd (hunk 2 above)
Comment (balderdash) [reply bin/dotd]: It just seems simpler to me, but others might disagree. It also conveniently identifies the latest passing test build without having to check slack, which could come in handy elsewhere.

Comment (ashercodeorg) [reply bin/dotd]: Done via `git push origin --delete #{branch_name}`.

**Comment by ashercodeorg** [comment]
Though longer term I'm fine with adopting a common `test-green` branch, in the short term I'm opposed to doing so because of the (at least perceived) increased risk of DTPing the wrong commit. This pipeline isn't well-tested, I'd just assume make sure it is well-oiled before introducing a common branch.

**Comment by ashercodeorg** [comment]
PTAL.

**Pull request merged**